src/Microsoft.Health.Fhir.TemplateManagement/ArtifactProviders/BlobTemplateProvider.cs

```diff
@@ -0,0 +1,99 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using DotLiquid;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Health.Fhir.Liquid.Converter.Utilities;
+
+namespace Microsoft.Health.Fhir.TemplateManagement.ArtifactProviders
+{
+    public class BlobTemplateProvider : IConvertDataTemplateProvider
+    {
+        private BlobContainerClient _blobContainerClient;
+
+        private readonly IMemoryCache _templateCache;
+
+        private readonly string _blobTemplateCacheKey = "cached-blob-templates";
+
+        private readonly int _segmentSize = 100;
+
+        public BlobTemplateProvider(BlobContainerClient blobContainerClient, IMemoryCache templateCache)
+        {
+            _blobContainerClient = blobContainerClient;
+            _templateCache = templateCache;
+        }
+
+        public async Task<List<Dictionary<string, Template>>> GetTemplateCollectionAsync(CancellationToken cancellationToken = default)
+        {
+            // read templates from cache if available
+            if (_templateCache.TryGetValue(_blobTemplateCacheKey, out var templateCache))
+            {
+                return (List<Dictionary<string, Template>>)templateCache;
+            }
+
+            var templateLookup = new Dictionary<string, Template>();
+            var templateNames = await ListBlobsFlatListing(_blobContainerClient, _segmentSize);
+
+            foreach (var templateName in templateNames)
+            {
+                var template = await DownloadBlobToStringAsync(_blobContainerClient, templateName);
+                var renamedTemplate = TrimBlobName(templateName);
+                Template liquidTemplate = TemplateUtility.ParseTemplate(renamedTemplate, template);
+                templateLookup.Add(renamedTemplate, liquidTemplate);
+            }
+
+            var templates = new List<Dictionary<string, Template>>
+            {
+                templateLookup,
+            };
+
+            _templateCache.Set(_blobTemplateCacheKey, templates);
```

Comment:
What is the policy for expiring entries in the cache?  How does that compare with the ACR policy?

Reply:
Right now for this PR the cache for storage templates won't expire. I can pass some cache options in. The ACR cache logic is pretty confusing and not sure we want to replicate it. There are long and short expiry times of 60 minutes and 20 minutes here: https://github.com/microsoft/FHIR-Converter/blob/baf921359046571421d26f52d8360d8b34045d76/src/Microsoft.Health.Fhir.TemplateManagement/Models/TemplateCollectionConfiguration.cs#L10

But also the template provider itself seems to only be cached for 5 minutes (so I think 5 minutes is the real value since the template provider gets re-initialized every 5 minutes and therefore anything in the 20 minute cache or 60 minute cache seems like it would be lost): https://github.com/microsoft/fhir-server/blob/536a1021d32226645f8e8338345ecdc90afea198/src/Microsoft.Health.Fhir.Core/Features/Operations/ConvertData/ContainerRegistryTemplateProvider.cs#L98C29-L98C51

I was thinking we could have that data plane operation we proposed for template settings locations(s) and template settings in general could be used to invalidate the cache, or set the cache time. So at the very least I should probably pass some cache options in.

Reply:
added options for caching

Reply:
Thanks Will, this could be something we incorporate into the new API design as well. Maybe we allow the user to bypass cache based on a header or param when specified.

Thank you for breakdown of the ACR policy.  What you shared mirrors my understanding of the code which was slightly confusing to me because one of the reasons shared for using ACR was to have a single hash for comparison to determine the cache should be invalidated.  So I was expecting we would be at least validating that the hash hasn't changed at the source with each request but that doesn't appear to be the case.